test/e2e/storage/testsuites/subpath.go

```diff
@@ -0,0 +1,746 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testsuites
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/drivers"
+	"k8s.io/kubernetes/test/e2e/storage/testpatterns"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var (
+	volumePath      = "/test-volume"
+	volumeName      = "test-volume"
+	probeVolumePath = "/probe-volume"
+	probeFilePath   = probeVolumePath + "/probe-file"
+	fileName        = "test-file"
+	retryDuration   = 20
+	mountImage      = imageutils.GetE2EImage(imageutils.Mounttest)
+)
+
+type subPathTestSuite struct {
+	tsInfo TestSuiteInfo
+}
+
+var _ TestSuite = &subPathTestSuite{}
+
+// InitSubPathTestSuite returns subPathTestSuite that implements TestSuite interface
+func InitSubPathTestSuite() TestSuite {
+	return &subPathTestSuite{
+		tsInfo: TestSuiteInfo{
+			name: "subPath",
+			testPatterns: []testpatterns.TestPattern{
+				testpatterns.DefaultFsInlineVolume,
+				testpatterns.DefaultFsPreprovisionedPV,
+				testpatterns.DefaultFsDynamicPV,
+			},
+		},
+	}
+}
+
+func (s *subPathTestSuite) getTestSuiteInfo() TestSuiteInfo {
+	return s.tsInfo
+}
+
+func (s *subPathTestSuite) skipUnsupportedTest(tp testpatterns.TestPattern, drv drivers.TestDriver) {
+}
+
+func createSubPathTestInput(tp testpatterns.TestPattern, tr subPathTestResource) subPathTestInput {
+	drv := tr.driver
+	di := drv.GetDriverInfo()
+	f := di.Framework
+	subPath := f.Namespace.Name
+	subPathDir := filepath.Join(volumePath, subPath)
+
+	return subPathTestInput{
+		f:                 f,
+		subPathDir:        subPathDir,
+		filePathInSubpath: filepath.Join(volumePath, fileName),
+		filePathInVolume:  filepath.Join(subPathDir, fileName),
+		volType:           tr.volType,
+		pod:               tr.pod,
+		volSource:         tr.genericVolumeTestResource.volSource,
+		roVol:             tr.roVolSource,
+	}
+}
+
+func (s *subPathTestSuite) execTest(drv drivers.TestDriver, tp testpatterns.TestPattern) {
+	Context(getTestNameStr(s, tp), func() {
+		var (
+			tr           subPathTestResource
```

Comment:
Can we use non-abbreviated variable names throughout?  Like "resource", "input", and "pattern"?  It's more verbose, but will be easier to read and understand.